Conan build packages on CI (#4590)

uilianries opened the pull request:
Hi again!

Related issues: #4590 and #4522 

This PR brings the opportunity to build Flatbuffers using multiple configuration on Linux (gcc and clang), OSX (clang) and Windows (msvc), and distribute on Bintray. 

After to introduce Conan recipe in Flatbuffers, any developer is able to create a package locally, copying all Flabuffers' code base and running Conan client.

But now, both Travis and Appveyor will start a complete Conan build only when a new release is created (tag) on this project. All other builds weren't affected, they are working as before.

A bunch of scripts were added to help these jobs. Let's explain in more details:

1) I just replaced the old code injection used [here](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-2b1d42f71f22b7ea0412d7602dec166fR24) to use a [cmake wrapper](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-f27818ade1d94739dd6d6f87fbadd78dR1). This is more clean and readable IMO.

2) Each binary package provided by Conan contains a specific configuration, based on settings + options (os x build_type x arch x compiler x [dynamic lib|static lib]). So, each Conan build on [Linux](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-354f30a63fb0907d4ad57269548329e3R107) and [Mac](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-354f30a63fb0907d4ad57269548329e3R121) could have 8 to 16 jobs and Windows have 3 by appveyor job. The [ConanMultiplePackage](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-d3233b2c48c826f17968ca90134653bfR78) is a helper to provide all jobs.

3) As said before, only when a new release is created, Conan will be started. This behavior will happen on [Travis](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-354f30a63fb0907d4ad57269548329e3R17) and [Appveyor](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-a3ca249bdb00ada6539de4f410a3c10eR18). Unfortunately, there is no option to switch tag on master on yaml, when running on Appveyor. So, I used such script to this.

4) To validate each new package, a test was introduced in conan/test_package, this include a tidy project just to make sure that there is header and library exported by the package. The small project includes a [conanfile](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-39b64f799cef8119375adb87053164d9R1) to import all package stuff and execute the test, a [cmake file](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-a8bf759f87ba1f78f44802b14fb4e5e6R1) to run conan_basic_setup and build the application, and finally a [cpp file](https://github.com/google/flatbuffers/compare/master...uilianries:feature/conan-ci#diff-97d2111886f70043ee9dbc062e167584R1) just to validate some header and lib linkage.

5) To be able to push on Bintray, a new account should be created. You could find a nice documentation on [conan docs](http://docs.conan.io/en/latest/uploading_packages/bintray/uploading_bintray.html). After to create a new Bintray account, you will need to fill two env vars on both CI services (Travis and Appveyor):
* CONAN_LOGIN_USERNAME: <your_bintray_username>
* CONAN_PASSWORD: <you_bintray_api_key>
By default, your packages will be uploaded to https://bintray.com/flatbuffers/conan, but you can change the name *flatbuffers* to any other by adding `CONAN_USERNAME` on your env vars. This variable works like the organization name, and `CONAN_LOGIN_USERNAME` works like the user allowed to push on the organization.

6) The PR will not start a Conan build to check all these stuff, however, I created feature branch, master and tag on same commit to validate:
* Travis
  * [master](https://travis-ci.org/uilianries/flatbuffers/builds/328856808)
  * [feature branch](https://travis-ci.org/uilianries/flatbuffers/builds/328857166)
  * [tag](https://travis-ci.org/uilianries/flatbuffers/builds/328856959)

* Appveyor
  * [master](https://ci.appveyor.com/project/uilianries/flatbuffers/build/1.0.49)
  * [tag](https://ci.appveyor.com/project/uilianries/flatbuffers/build/1.0.50)
  * on Appveyor, feature branches are not built

As you can see, this PR is a huge improvement for Conan stuff, so feel free ask any question!

Also, I adding some Conan experts to help on review
/cc @memsharded @lasote @solvingj

Regards!

memsharded commented on conan/appveyor/install.py:
```diff
@@ -0,0 +1,25 @@
+#!/usr/bin/env python
```
IMHO, this is too lengthy. I am not sure I get all the abstractions as useful, I would do just:

```python
import os
if os.getenv("APPVEYOR_REPO_TAG") != "true":
    print("Skip...")
else:
    os.system("pip install conan conan-package-tools")
```

uilianries replied:
seems reasonable. I'll adopt your solution.

memsharded commented on conan/build.py:
```diff
@@ -0,0 +1,93 @@
+
+if __name__ == "__main__":
+    name, version, username, channel = get_env_vars()
+    reference = "{0}/{1}".format(name, version)
```
I am not sure if I get why it is necessary to get the package name & version from the recipe, they are already there, no need to extract them to pass them back.

uilianries replied:
how could take both values to set `CONAN_REFERENCE`? Is there some trick using the API?

solvingj replied:
Indeed, I would also like clarification.

memsharded replied:
But I'd say in most cases you don't need the reference, if name/version are defined in the recipe, and user/channel are env-vars.

memsharded commented on conan/test_package/conanfile.py:
```diff
@@ -0,0 +1,25 @@
+        cmake.build()
+
+    def test(self):
+        with tools.environment_append(RunEnvironment(self).vars):
```
I have already seen this in another package. @uilianries it seems it could be nice to have it automated in some conan tool, to simplify the ``test()`` method

uilianries replied:
thanks to @solvingj. I could merge into conan on a next PR

memsharded commented on conanfile.py:
```diff
@@ -45,6 +43,13 @@ def package(self):
         self.copy(pattern="flathash*", dst="bin", src="bin")
+        if self.settings.os == "Windows" and self.options.shared:
+            if self.settings.compiler == "Visual Studio":
+                shutil.move(os.path.join(self.package_folder, "lib", "%s.dll" % self.name),
```
Why these extra ``shutil.move``? Why not standard ``self.copy()``?
Also, why necessary if there is already a ``cmake.install()``?

uilianries replied:
flatbuffers installs runtime libraries on `lib` folder. After to install all libraries, I moved from lib to bin. Otherwise, will be necessary a workaround on `package_info`.

memsharded replied:
Oh, I see, thanks for the explanation, it is ok then.

memsharded commented on conan/build.py:
```diff
@@ -0,0 +1,93 @@
+
+def add_visual_build(builder, version, arch, build_type, runtime, shared):
+    settings = {'compiler.version': version, 'arch': arch, 'build_type': build_type,
+                'compiler.runtime': '%s%s' % (runtime, ('d' if build_type == "Debug" else '')),
```
Runtime might be simplified if doing only MD/MDd, as now conan can dynamically adjust it to release/debug according to ``build_type``

uilianries replied:
The point here is that conan-package-tools doesn't solve x86_64 for Visual Studio 10. Would be nice to solve all this stuff directly from package tools.

uilianries replied:
My fault! I just revised package tools and I found vs10_x86_64_enabled. I'll remove this block.

aardappel commented:
Thanks for all the work! Some notes:

- Not super happy with the large amounts of Python being introduced, because when any of this breaks, I am not sure I'll have time to figure out how to fix it. Is this how you normally build things with Conan? Why is this necessary, and why can't Conan do more of this out of the box?
- As for binary build matrix, what is most useful is that people can rely on a downloadable `flatc` executable, which can be just x64 release mode, and does't need all other combinations. Pre-built binaries for libs is not as useful, since FlatBuffers is small and mostly intended to be statically linked, and in most use-cases is actually header-only. But I suppose it doesn't hurt to have them, if you feel strongly about it.
- In particular, it would be great if this causes `flatc` for the 3 platforms to be available at a predictable url, since recently e.g. @yfinkelstein has been setting up automatic packaging for Java/Maven, where being able to pack in those binaries would be super useful. Similarly, we may want to supply these binaries for other languages / packaging systems (e.g. npm, pip..).

memsharded commented:
Thanks for the feedback!

> Not super happy with the large amounts of Python being introduced, because when any of this breaks, I am not sure I'll have time to figure out how to fix it. Is this how you normally build things with Conan? 

Yes, I agree. I think an effort to reduce its complexity has to be done. Many times it is not necessary such code, for example, I am using just this script: https://github.com/memsharded/conan-zmq/blob/4.2.0/build.py to build libzmq for tenths of the most typical configurations in all OSs. I will try to collaborate to improve this

> As for binary build matrix, what is most useful is that people can rely on a downloadable flatc executable, which can be just x64 release mode, and does't need all other combinations. Pre-built binaries for libs is not as useful, since FlatBuffers is small and mostly intended to be statically linked, and in most use-cases is actually header-only. But I suppose it doesn't hurt to have them, if you feel strongly about it.

Yes, I want to do the same for a protobuf package, to split the protoc into its own package, that will use static linkage, and just the OS and arch to be defined. I'd say it is an identical case. Pre-built binaries for libs might be well skipped with ``build_policy="missing"`` in the recipe, but IMHO it is a good validation to have the binaries there, it is a proof that such platform/configuration is supported, so you don't need to find out it doesn't build when building from sources. 

> In particular, it would be great if this causes flatc for the 3 platforms to be available at a predictable url
Once uploaded to bintray, the file URL can be obtained, and it is predictable given the configuration ID, but this is a sha1 hash, so it is not human-friendly in the sense that you type some ``url/path/to/release/2.3/flatc`` and you can get it.

aardappel commented:
Ah.. for scripts (other language package managers) is there a way to programmatically figure out this URL?

For humans, I can link to these URLs from the releases page, at least it will save me building them myself.

What is the state of this PR? Would you prefer to merge it and improve later?

aardappel commented:
We have a repo on bintray: https://bintray.com/aardappel/flatbuffers, and travis/appveyor environment keys have been set.

memsharded commented:
I am just checking and this would be an example: https://bintray.com/conan-community/conan/7z_installer%3Aconan/1.0%3Astable#files/conan%2F7z_installer%2F1.0%2Fstable%2Fpackage%2F3475bd55b91ae904ac96fde0f106a136ab951a5e. So it might be doable to automate, but the inconvenient would be that conan actually creates a .tgz file with all the package contents there.

I think for distributing ``flatc`` it might be better to just upload the files to a generic repo to get something shorter and easier to automate. Should be relatively easy to automate in CI.

aardappel commented:
You're saying I should set up additional CI because bintray has hard to use URLs? Why should we integrate with bintray at all then?

memsharded commented:
Not additional CI job, just an extra upload at the end of the process, uploading ``flatc`` to bintray too, but to an easier URL. In any case, I am not sure, I need to investigate, maybe bintray has some other mechanism to simplify URLs, I will have a look.

Do you think that distributing ``flatc`` inside a ``.tgz`` file would be an issue for users or maybe not?

solvingj commented:
For what it's worth, there's an alternate URL format you can use with Bintray for any repo type at the `dl.bintray.com` domain: 

An example for FMT: 
https://dl.bintray.com/bincrafters/public-conan/bincrafters/fmt/4.0.0/stable/package/e0a02d496bbb652b6295152dfce0d3937acc0b56/conan_package.tgz

This still requires you construct the URL from a bunch of relevant variables, and in the case of Conan packages this includes the SHA.  It's possible to query and grep the SHA for a local binary using `conan info` command, but maybe in this case there's an easier way @memsharded can think of (if doing so is even necessary).

uilianries commented:
Hi! 

I just updated this PR to use less code as necessary. There is no different behavior, only less code.

Also, I updated to upload all packages on *bintray.com/aardappel/flatbuffers*, but could be customized by `CONAN_UPLOAD`.

Regards!

aardappel commented:
@memsharded .tgz should not be a problem.. even on Windows most archivers support it.

uilianries commented:
Hi again!

I just updated this PR to use new compiler versions.

You could check the CI status here:
https://travis-ci.org/uilianries/flatbuffers/builds/415252608
https://ci.appveyor.com/project/uilianries/flatbuffers/build/1.0.62

Also, I created a tag to start full Travis build:
https://travis-ci.org/uilianries/flatbuffers/builds/415252831

What is needed to continue this PR?

Croydon commented on .travis.yml:
```diff
@@ -73,7 +73,7 @@ matrix:
       script:
       - cmake -DCMAKE_BUILD_TYPE=$BUILD_TYPE . && make && make test
-      - if [ "$CONAN" == "true" ] && [ "$TRAVIS_OS_NAME" == "linux" ]; then sudo pip install conan && conan create . flatbuffers/testing -s build_type=$BUILD_TYPE; fi
+      - if [ "$CONAN" == "true" ] && [ "$TRAVIS_OS_NAME" == "linux" ]; then sudo pip install conan && conan create . aardappel/testing -s build_type=$BUILD_TYPE -tf conan/test_package; fi
```
Wouldn't it better to use google as the username? This would create some consistency across Google libraries. E.g. I have suggested to use `@google` for the benchmark library as well https://github.com/google/benchmark/pull/647

uilianries replied:
Done!

Croydon commented on .travis.yml:
```diff
@@ -62,6 +85,37 @@ matrix:
+      osx_image: xcode9
+      env: CONAN_APPLE_CLANG_VERSIONS=9.0
+    - <<: *conan-osx
+      osx_image: xcode9.3beta
```
Is there a specific reason for beta?

uilianries replied:
no, I just copied from zlib recipe. I'll update

uilianries replied:
Done!

aardappel commented:
@memsharded you LGTM as well?

Not that familiar with Conan, so when y'all are LGTM I can merge.

uilianries commented:
Adding @lasote and @danimtb. They are Conan official members.

@memsharded is on vacation :tropical_drink: :beach_umbrella:

danimtb commented:
@uilianries @aardappel LGTM as well. Great job!

aardappel commented:
Thanks everyone!

uilianries commented:
@aardappel you need to set `CONAN_PASSWORD` on CI env vars (Travis and AppVeyor). During the CI execution, Conan will upload all packages using your Bintray account. Your username and repo address are specified in Conan scripts.

The password is provided on your Bintray account, the API key.
To get your API key, you will need to access https://bintray.com/profile/edit and click on "API Key"

For example:

![bintray](https://screenshotscdn.firefoxusercontent.com/images/06be757a-4726-44f3-8a2c-1c4916382773.png)

References:
https://docs.conan.io/en/latest/uploading_packages/bintray/uploading_bintray.html

aardappel commented:
@uilianries I had done this already a while ago (see above), and checking in travis shows that it is set. Do these get invalidated every so often?

uilianries commented:
Nope. I just forgot to read the comments above :)

On Thu, Aug 16, 2018, 7:37 PM Wouter van Oortmerssen <
notifications@github.com> wrote:

> @uilianries <https://github.com/uilianries> I had done this already a
> while ago (see above), and checking in travis shows that it is set. Do
> these get invalidated every so often?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/google/flatbuffers/pull/4594#issuecomment-413705112>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEpQHfZb50EYUB6z7PpTeyjNz-UU1KScks5uRfQUgaJpZM4ReNlQ>
> .
>

Croydon commented:
Nice work anyone.

Now it would be nice if we can get within some reasonable time frame a new release of flatbuffers, so we have actually packages on Bintray

plus as a follow up the inclusion in conan-center

aardappel commented:
@Croydon we can probably do a release sometime soon. Though maybe worth it to manually trigger one for 1.9?

wpalfi commented:
Please trigger one for 1.9!

Croydon commented:
Depending on how soon a new release would happen anyway it might not be worth the effort I guess. It would require to create a new branch, back port the Conan recipe and the CI additions and adjusting them for an one-time run. (since we don't want and can't have a second 1.9 git tag which would normally trigger the package building and uploading).

However, if anyone wants to take the time for this then it would be fine of course.

aardappel commented:
I've made a note to get 1.10 going.. but can't make promises, it may be a few weeks, since I am going on a trip.

Croydon commented:
@aardappel Awesome! 

Could you now please request the inclusion in conan-center?

Go to https://bintray.com/aardappel/flatbuffers/flatbuffers%3Agoogle and click the "Add to Conan Center" button.

aardappel commented:
@Croydon Can't find such a button.. do you mean "Set Me Up!" ? :)

Croydon commented:
@aardappel If you are logged in you should see in the flatbuffer package overview something like this:

![conancenter](https://user-images.githubusercontent.com/1593194/46443957-1fe27700-c770-11e8-841b-583b12c2174d.PNG)

aardappel commented:
Doh! Done.

uilianries commented:
Awesome! 🍻 

Bintray's page there are fields related to Conan package. Unfortunately those field are not filled by Conan, they need manually update.

Below I filled all requested fields in red color. Also you could use Flatbuffers' logo and some tags. To edit those fields, just click on "Edit" button on the top level.

[Catch2](https://bintray.com/bincrafters/public-conan/catch2%3Abincrafters) could be a nice example if you are interested.

![screenshot_2018-10-03 flatbuffers google](https://user-images.githubusercontent.com/4870173/46445570-eef14a80-c74d-11e8-9777-8765917f877b.png)

Croydon commented:
It is now available on conan-center 🎉 

https://bintray.com/conan/conan-center?filterByPkgName=flatbuffers%3Agoogle
